### PR TITLE
REST API docs via OpenAPI

### PR DESCRIPTION
**Pull request by bourque**
# Change Summary

## Overview

This PR adds a new documentation page that describes our REST API. It is mostly auto-generated using OpenAPI, though I added some customized examples.

## New Dependencies
- `sphinxcontrib-openapi` -- `sphinx` extension to support auto-generated OpenAPI docs

## New Files

- `docs/source/rest-api/index.rst`
    - Index page for the REST API docs
- `docs/source/rest-api/openapi.yml` 
    - OpenAPI configuration for our SDC REST API

## Updated Files

- `docs/source/conf.py`
    - Added the `sphinxcontrib-openapi` extension
- `docs/source/index.rst`
    - Added link to the REST API docs
- `pyproject.toml` and `poetry.lock`
    - Added `sphinxcontrib-openapi` dependency

### CODE REVIEW

File: docs/source/conf.py
Comment (bourque) [inline docs/source/conf.py:44]: Just putting some things in alphabetical order here

File: docs/source/rest-api/index.rst
Comment (bourque) [inline docs/source/rest-api/index.rst]: Unfortunately I had to manually create some examples here. Currently the `sphinxcontrib-openapi` package does not automatically generate examples listed in the `openapi.yml` configuration successfully, there are some known issues with it. I guess the upside is that we will have finer control over what we want our users to see.

File: docs/source/index.rst
Comment (maxinelasp) [inline docs/source/index.rst]:
```suggestion
   api/index
```
I'd suggest leaving "rest" out of it, but that's not a super strong preference

File: docs/source/rest-api/index.rst
Comment (maxinelasp) [inline docs/source/rest-api/index.rst]: Ditto for just "API" or maybe "File Access API"

File: docs/source/rest-api/index.rst
Comment (maxinelasp) [inline docs/source/rest-api/index.rst]: Since this just confused me, could you add the dev API url (`https://api.dev.imap-mission.com`) so in the near future instrument teams can use this as a reference?

**Review by maxinelasp — APPROVED** [review]
Looks great! Thanks for adding this!

File: docs/source/rest-api/index.rst
Comment (tech3371) [inline docs/source/rest-api/index.rst]:
```suggestion
   curl -X GET -H "Accept: application/json" https://api.dev.imap-mission.com/upload?filename=imap_codice_l1a_lo_20260101_20260102_v01-01.cdf
```

File: docs/source/index.rst
Comment (greglucas) [reply docs/source/index.rst]: Data API? Or some other qualifier if we want to be more explicit what it is an API for. I agree, REST is how it functions, but not descriptive.

File: docs/source/rest-api/index.rst
Comment (greglucas) [inline docs/source/rest-api/index.rst]:
```suggestion
   curl -X GET -H "Accept: application/json" https://api.dev.imap-mission.com/upload?filename=imap_codice_l1a_lo_20260101_20260102_v01-01.cdf
```

File: docs/source/rest-api/index.rst
Comment (tech3371) [inline docs/source/rest-api/index.rst]:
```suggestion
   {"statusCode": 200, "body": "https://sds-data.s3.amazonaws.com/imap/codice/l1a/2026/01/imap_codice_l1a_lo_20260101_20260102_v01-01.cdf?<long-credentials-string>"}
```

File: docs/source/rest-api/index.rst
Comment (greglucas) [reply docs/source/rest-api/index.rst]: We will want to move this onto the SDC's main website once that is up anyways, so I wouldn't spend too much time on these inconsistencies yet.

File: docs/source/rest-api/index.rst
Comment (greglucas) [inline docs/source/rest-api/index.rst]: I am not sure we can ever get the 200 success anymore? i.e. a successful response is a 302.

File: docs/source/rest-api/index.rst
Comment (greglucas) [inline docs/source/rest-api/index.rst]:
```suggestion
   curl -X GET -H "Accept: application/json" https://api.dev.imap-mission.com/download?s3_uri=s3://sds-data/imap/codice/l1a/2026/01/imap_codice_l1a_lo_20260101_20260102_v01-01
```
All of these are query parameters currently I believe and not path parameters.

Comment (tech3371) [inline docs/source/rest-api/index.rst]:
```suggestion
   curl -X GET -H "Accept: application/json" https://api.dev.imap-mission.com/download?s3_uri=s3://sds-data/imap/codice/l1a/2026/01/imap_codice_l1a_lo_20260101_20260102_v01-01.cdf
```

File: docs/source/rest-api/openapi.yml
Comment (greglucas) [inline docs/source/rest-api/openapi.yml]: Do we need "*"? I would say we should just leave the parameter off the request to make it a wildcard instead.

File: docs/source/rest-api/openapi.yml
Comment (greglucas) [inline docs/source/rest-api/openapi.yml]: I think @sdhoyt may be updating the query APIs, so it would be good for him to let us know what these values are or update these later, because I think we'll want startdate and enddate (or some other spelling of that) to make a range for the query.

File: docs/source/rest-api/index.rst
Comment (greglucas) [inline docs/source/rest-api/index.rst]:
```suggestion
   curl -X GET -H "Accept: application/json" https://api.dev.imap-mission.com/query?level=l0&instrument=mag&date=20230112&version=*&extension=pkts
```
I believe we are getting rid of mission as an option. Do we have a list of what options we allow somewhere as well?

File: docs/source/rest-api/openapi.yml
Comment (tech3371) [inline docs/source/rest-api/openapi.yml]: should we change to this since it has parameter information right below it? It could confuse `filename` with endpoint.
```suggestion
  '/upload':
```

File: docs/source/rest-api/openapi.yml
Comment (tech3371) [inline docs/source/rest-api/openapi.yml]: same here
```suggestion
  '/download':
```

**Review by greglucas — COMMENTED** [review]
This looks great!!

I think all of our API is going through query parameters right now. Do we want to change any of those to path parameters on the backend?

I added quite a few comments on the content of the API/server that are unrelated to this PR, so feel free to ignore them and just consider them left there for us to think about updating on the backend.

**Review by tech3371 — COMMENTED** [review]
It looks great! Thank you! I had few minor corrections.

File: docs/source/index.rst
Comment (bourque) [reply docs/source/index.rst]: I like "Data API" or perhaps "Data Access API"

File: docs/source/rest-api/index.rst
Comment (bourque) [reply docs/source/rest-api/index.rst]: Good catch!

File: docs/source/rest-api/index.rst
Comment (bourque) [reply docs/source/rest-api/index.rst]: I think these API docs themselves could serve as the list of available options. Unless you are thinking we should list these somewhere else too?

Comment (greglucas) [reply docs/source/rest-api/index.rst]: Nope, I completely agree with this being the reference point! Just wanted to make sure that @sdhoyt is also aware of what options you may be submitting and will be defining the interaction with the DB.

File: docs/source/data-access-api/index.rst
Comment (sdhoyt) [inline docs/source/data-access-api/index.rst]: Do want to add the new response from the RDS query PR or wait until that's merged?:

```
{
"statusCode": 400,
  "body": "<param> is not a valid query parameter. Valid query parameters are: <valid parameters>"),
    "headers": {
         "Content-Type": "application/json",
          "Access-Control-Allow-Origin": "*", 
    },
}
```

**Review by sdhoyt — APPROVED** [review]
This looks great!

File: docs/source/rest-api/openapi.yml
Comment (sdhoyt) [reply docs/source/rest-api/openapi.yml]: These are what I'm currently considering valid, but I imagine this list will change a bit as time goes on:
- file_path
- instrument
- data_level
- descriptor
- start_date
- end_date
- version
- extension

File: docs/source/data-access-api/index.rst
Comment (bourque) [reply docs/source/data-access-api/index.rst]: Thanks for this! I can add it to this PR, assuming it is not going to change anytime soon.

**Comment by bourque** [comment]
@greglucas @sdhoyt I pushed some changes that hopefully reflect what we are likely to merge in with #317 and https://github.com/IMAP-Science-Operations-Center/sds-data-manager/pull/228. There will still likely be some tweaks we will want to make, but I think this is a good starting point. We should discuss this at our next meeting.

File: docs/source/data-access-api/openapi.yml
Comment (greglucas) [inline docs/source/data-access-api/openapi.yml]:
```suggestion
          description: Search for all files with a start date on or after this time, in the format ``YYYYMMDD``.
```
and for the end_date:
`description: Search for all files with a start date on or before this time, in the format ``YYYYMMDD``.`

Do we want to follow the date hint?
https://swagger.io/docs/specification/data-models/data-types/
> date – full-date notation as defined by [RFC 3339, section 5.6](https://tools.ietf.org/html/rfc3339#section-5.6), for example, 2017-07-21

Looks like this date isn't the suggested with a "-" if we care about that.

File: docs/source/data-access-api/openapi.yml
Comment (greglucas) [inline docs/source/data-access-api/openapi.yml]: I think this is the only one left of "*", did you want to remove this too?
```suggestion
            The file extension of interest (e.g. ``pkts``). Supported extensions include
            ``pkts`` and ``cdf``.
```

**Review by greglucas — APPROVED** [review]
Looks good to me.

File: docs/source/data-access-api/openapi.yml
Comment (bourque) [reply docs/source/data-access-api/openapi.yml]: Good question. I think I will punt on this for now, but bring it up for discussion when we go over the API.

File: docs/source/data-access-api/openapi.yml
Comment (bourque) [reply docs/source/data-access-api/openapi.yml]: Good catch!

**Pull request merged**
